### PR TITLE
DexEx IDC: Batch Deletion (Multithreading)

### PR DESCRIPTION
**Pull request by mohsha-msft**
Solution to [this problem](https://github.com/wastore/xstorehackathon/blob/main/BatchDelete.md)
Please refer [README.md](https://github.com/mohsha-msft/xstorehackathon/blob/dexex-idc/batch-delete/batch-deletion/README.md) for more information.

### CODE REVIEW

File: batch-deletion/main.py [hunk 1]
```diff
@@ -0,0 +1,167 @@
...
+    global MaxThreadCount
+
+    if cli_options['connection_string']:
+        azure.set_connection_string(cli_options['connection_string'])
+    else:
+        conn_str = os.getenv("AZURE_STORAGE_CONNECTION_STRING")
+        azure.set_connection_string(conn_str)
+
+    if cli_options['parallel_factor']:
+        MaxThreadCount = int(cli_options['parallel_factor'])
```
Comment (tasherif-msft) [inline batch-deletion/main.py:120]: nit: The cli option `parallel_factor` is a little misleading. If I did not want to run in parallel I might just naturally pass 0 (trying to only run 1 thread)

File: batch-deletion/main.py [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+    # Start the thread to report the usage info
+    report_thread = threading.Thread(target=delete_batch)
+    report_thread.daemon = True
+    report_thread.start()
+
+    # Start N threads to iterate the directory and list recursively
+    for i in range(MaxThreadCount):
+        # print("Starting child thread " + str(i))
+        t = threading.Thread(target=thread_handler, args=[i])
+        t.daemon = True
```
Comment (tasherif-msft) [inline batch-deletion/main.py:145]: Is there a reason why you're instantiating the threads as daemon threads? You're joining them later on - so they're technically acting as non-daemon threads.

**Review by tasherif-msft — APPROVED** [review]
Besides the two nits, this is excellent work :) great work on thinking of the edge cases and locks!

File: batch-deletion/main.py (hunk 1 above)
Comment (mohsha-msft) [reply batch-deletion/main.py:120]: If `parallel_factor` is not passed, we use default `MaxThreadCount=8`. The whole purpose of this Batch Deletion code we wrote was to implement it using Multithreading. If multithreading is not required, user can pass `parallel_factor=1`

File: batch-deletion/main.py (hunk 2 above)
Comment (mohsha-msft) [reply batch-deletion/main.py:145]: Ohh we didn't know about that. Never used Python before this hackathon actually :)

**Pull request closed without merging**